### ISSUE TITLE
Metal command buffer lstm failed: 5

### ISSUE THREAD

**Issue by P3R53U5M** [nanoporetech/dorado#237]
I'm currently basecalling flongle data from R10 pores with kit 14 chemistry on an M1 macbook 16 GB ram with the sup model.

dorado basecaller ./dna_r10.4.1_e8.2_400bps_sup@v4.2.0/ ./ > ./001_sup_v4.2.0.bam 

I just got the following error messages repeatedly, but the basecaller keeps running:

[warning] Metal command buffer lstm failed: 5, try #0
[2023-06-13 08:15:42.844] [warning] Command buffer error code: 1

What does this error mean?  I can't find documentation.  Is it ok to use this basecalled data with this error - or should I restart?

**Comment by tijyojwad** [comment]
Hello,

This error refers to an internal error code from Metal. I was able to reproduce this locally by interrupting a run and then rerunning basecalling on the same device.

However, when I run dorado again after a successful basecalling run, I don't see the error. Perhaps some internal state gets reset. This is not the ideal solution, but you could try to clean the state by basecalling just 1 read, and then trying the whole dataset again
```
dorado basecaller ./dna_r10.4.1_e8.2_400bps_sup@v4.2.0/ ./ --max-reads 1 > /dev/null
dorado basecaller ./dna_r10.4.1_e8.2_400bps_sup@v4.2.0/ ./ > ./001_sup_v4.2.0.bam
```

**Comment by P3R53U5M** [comment]
Thanks for the response.  So is it ok to use the data from basecalls with these errors?  O should I re-run it with your suggestion?

It takes me over a day to run this so I’d rather not scrap the data.

Perseus

**Comment by StuartAbercrombie** [comment]
This is a tricky one.

The more common failure scenario is a command buffer timeout, but these internal errors are rarer and usually have no obvious cause.  I've most often seen it when there are multiple GPU tasks running concurrently.  There is code that will retry command buffers, mainly intended to deal with timeout errors, but I don't think it will have helped here.

It is hard to know what exactly happened in this case, but it looks to me that the lstm kernel failing in this way will just log this message and carry on, probably with bogus results (we should do better on this).  If it was a one-off message then I'd expect the results to be OK except for one batch of read chunks.

In general running sup on an M1 is going to be a challenge.  Hac would be less likely to run into bad GPU states.

**Comment by mattloose** [comment]
Hi,

I'm seeing this error a lot when trying to run HAC basecalling on dorado 0.3.1 on an M2 Max.

Is there anything that can be done by tweaking batch sizes etc to help resolve it?

Thanks

**Comment by StuartAbercrombie** [comment]
The auto batch size on M2 Max should be the optimal one, but we can check that (I haven't personally testing on that chip, and it's possible GPU core count detection is not working correctly).  If you run with the -v option, it will log the retrieved GPU core count and batch size.  For an M2 Max it should report something like 38 cores and a batch size of this multiplied by 48 (there are variants with different GPU core counts).

All that said, I think the batch size if probably not having a large effect on the command buffer failure situation.  Just to check, you are still getting the error code 1?  Or are there timeouts reported now?

**Comment by mattloose** [comment]
THanks for coming back to me!

Running with -v gives:

[2023-07-03 09:11:10.836] [info] > Creating basecall pipeline
[2023-07-03 09:11:10.912] [debug] Retrieved GPU core count of 40 from IO Registry
[2023-07-03 09:11:11.281] [debug] Retrieved CPU performance core count of 8 from sysctl
[2023-07-03 09:11:11.282] [debug] - set batch size to 1920
[2023-07-03 09:11:12.500] [debug] Metal command buffer convolutions: 264.52176 ms
[2023-07-03 09:11:19.755] [debug] Metal command buffer lstm: 7253.8 ms
[2023-07-03 09:11:20.880] [warning] Metal command buffer linear/scan/softmax failed: 5, try #0
[2023-07-03 09:11:20.880] [warning] Command buffer error code: 1 (Internal Error (0000000e:Internal Error))
[2023-07-03 09:11:21.497] [debug] Metal command buffer convolutions: 269.23587 ms
[2023-07-03 09:11:28.752] [debug] Metal command buffer lstm: 7253.6216 ms
[2023-07-03 09:11:29.335] [warning] Metal command buffer linear/scan/softmax failed: 5, try #1
[2023-07-03 09:11:29.335] [warning] Command buffer error code: 1 (Internal Error (0000000e:Internal Error))

So it looks like it is misscounting the number of GPU cores?

**Comment by StuartAbercrombie** [comment]
Yes, that is a surprising outcome.  I don't have an exhaustive list of M2 Max configurations, but looking at M2 Max-based models currently available (Studio/MacBook Pro) the largest GPU core count advertised is 38.  I wasn't aware of a 40 GPU core M2 Max variant.  If this chip doesn't really have 40 GPU cores, it is quite strange that the IO Registry would say otherwise.  It could be that we simply cannot rely on this registry entry.

The time taken for executing the LSTM convolutions is also higher than expected, which could be down to an incorrect batch size for the chip.

As I say, these internal errors are hard to root cause because Metal doesn't tell us anything: the description of the error we retrieve is the "0000000e:Internal Error" part of what is logged.  Possibly it relates to GPU work taking too long (although usually this produces a different type of error), which might in turn relate to an incorrect batch size.  (There is discussion here of this type of error https://developer.apple.com/forums/thread/693849, without much resolution.)

I'm not sure how best to debug the GPU core count retrieval.  If you run this, you should get some relevant numbers we can compare:

`ioreg -l | grep num_cores`

We can try a batch size that should allow execution of LSTM work in a lower amount of time on the assumption that the GPU core count is < 40.  A safe value for all M2 Max configurations would be based on a GPU core count of 30, so I suggest trying -b 1440.  It would be interesting to see what it reports for the "Metal command buffer lstm" line with that set.

Finally, is anything else using the GPU for significant work while this is running?

**Comment by iiSeymour** [comment]
The documentation for `GPUConfigurationVariable` seems very sparse but it looks like `core_mask_list` is array of bit masks for `num_mgpus`: 

```python
>>> core_mask_list = (1023, 511, 511, 1023)
>>> sum(v.bit_count() for v in core_mask_list)
38
```

**Comment by iiSeymour** [comment]
@mattloose what do you from `system_profiler SPDisplaysDataType`?

**Comment by mattloose** [comment]
system_profiler SPDisplaysDataType
Graphics/Displays:

    Apple M2 Max:

      Chipset Model: Apple M2 Max
      Type: GPU
      Bus: Built-In
      Total Number of Cores: 38
      Vendor: Apple (0x106b)
      Metal Support: Metal 3
      Displays:
        SAMSUNG:
          Resolution: 3840 x 2160 (2160p/4K UHD 1 - Ultra High Definition)
          UI Looks like: 3840 x 2160 @ 60.00Hz
          Main Display: Yes
          Mirror: Off
          Online: Yes
          Rotation: Supported

**Comment by StuartAbercrombie** [comment]
That is very illuminating.  It does indeed look like we are getting the wrong batch size because we're not applying core_mask_list, and this leads to the internal error rather than the timeout report we usually see when the kernels take too long to complete.  On my M1 Pro, for example, I get core_mask_list (255,255) and we run as expected with 16 cores.

I'm looking at modifying the core querying code to account for core_mask_list.

**Comment by StuartAbercrombie** [comment]
I should say that in the interim you can manually set the optimum batch size for this M2 Max system, which would be 1824.

**Comment by iiSeymour** [comment]
@StuartAbercrombie has fixed this in 26be1a017baf7131a82271ebb624a484135d29ba 

@mattloose can I check one more thing `ioreg -l | grep gpu-core-count`.

**Comment by mattloose** [comment]
Sure:

Output is:

    | |   |   |   "gpu-core-count" = 38
